File: liste-envies-war/src/main/java/fr/desaintsteban/liste/envies/util/EncodeUtils.java
package fr.desaintsteban.liste.envies.util;

import java.util.Base64;
import java.nio.charset.StandardCharsets;

/**
 * Encode en base64 les textes pour empécher de les lires dans la base de donnée
 */
public class EncodeUtils {

    public static String encode(String string, boolean encode) {
        return encode ? encode(string) : string;
    }

    public static String encode(String string) {
        if (string != null) {
            byte[] message = string.getBytes(StandardCharsets.UTF_8);
            String encoded = Base64.getEncoder().encodeToString(message);
            return encoded;
        }
        return null;
    }

    public static String decode(String string, boolean decode) {
        return decode ? decode(string) : string;
    }

    public static String decode(String string) {
        if (string != null) {
            byte[] decoded = Base64.getDecoder().decode(string);
            return new String(decoded, StandardCharsets.UTF_8);
        }
        return null;
    }
}

File: liste-envies-back/src/common/utils/encode.utils.ts
/**
 * Encode en base64 les textes pour empêcher de les lire dans la base de données
 * Port de EncodeUtils.java
 */
export class EncodeUtils {
  /**
   * Encode une chaîne en Base64
   * @param str - La chaîne à encoder
   * @param shouldEncode - Si false, retourne la chaîne non encodée
   * @returns La chaîne encodée ou null
   */
  static encode(
    str: string | null | undefined,
    shouldEncode = true,
  ): string | null {
    if (!shouldEncode) {
      return str || null;
    }
    if (str != null && str !== undefined) {
      return Buffer.from(str, 'utf-8').toString('base64');
    }
    return null;
  }

  /**
   * Décode une chaîne depuis Base64
   * @param str - La chaîne encodée à décoder
   * @param shouldDecode - Si false, retourne la chaîne non décodée
   * @returns La chaîne décodée ou null
   */
  static decode(
    str: string | null | undefined,
    shouldDecode = true,
  ): string | null {
    if (!shouldDecode) {
      return str || null;
    }
    if (str != null && str !== undefined) {
      try {
        return Buffer.from(str, 'base64').toString('utf-8');
      } catch (error) {
        // En cas d'erreur de décodage, retourner la chaîne originale
        return str;
      }
    }
    return null;
  }
}
